docs(contributing): update guide (#3827)
algolia/instantsearch.js: CONTRIBUTING.md

Another element that will help us go faster at solving the issue is to provide a reproducible test case. We often recommend to [use this CodeSandbox template](https://codesandbox.io/s/github/algolia/instantsearch-templates/tree/master/src/InstantSearch.js).

## The code contribution process

InstantSearch.js was first developed in JavaScript and is now slowly being migrated to TypeScript.

For any code contribution, you need to:

- Fork and clone the project
- Create a new branch for what you want to solve (fix/_issue-number_, feat/_name-of-the-feature_)
- Make your changes
- Open a pull request

Depending on what you're working on, you might consider different [base branches](#branch-organization).

Then:

- Peer review of the pull request (by at least one of the core contributors)
- Automatic checks ([tests](#tests), [commits](#commit-conventions), [linters](#linting))
- When everything is green, your contribution is merged 🚀

After you create a pull request, a bot will comment with a link to a development version of the website.

> You will find there a playground for the widgets: https://**netlify-deploy-url**/stories

Reviewer: ```suggestion
You will find a playground for the widgets: https://**netlify-deploy-url**/stories
```

In most cases, we use the following types:

- `fix`: for any resolution of an issue (identified or not)
- `feat`: for any new feature
- `refactor`: for any code change that neither adds a feature nor fixes an issue

> - `chore`: for anything that is not related to the library itself (doc, tooling)

Reviewer: we should probably mention `docs` separately, since this what you naturally  did in this commit message even

Finally, if your work is based on an issue on GitHub, please add in the body of the commit message "fix #1234" if it solves the issue #1234 (read "[Closing issues using keywords](https://help.github.com/en/articles/closing-issues-using-keywords)").

Some examples of valid commit messages (used as first lines):

> - fix(searchbox): increase magnifying glass size
> - chore(deps): update dependency rollup-plugin-babel to v3.0.7
> - fix(connectRefinementList): set default value for limit
> - chore: reword contributions guides

## Branch organization

The project is based on the classic GitHub flow:

- `develop` for the current version being worked on – Pull requests for bugs and feature related to the current major version should be created against this branch
- `master` for the latest stable version – we usually don't make pull requests on this branch
- `vX` for each major version (`X` being a number) – Pull requests for critical bug fixes should be created against this branch

Most of the time, your pull requests should target the `develop` branch, except if you're fixing a critical bug on an older version.

> _Note that no new features will be developed on the `vX` branches._

Reviewer: ```suggestion
_Note that no new features will be developed or backported for the `vX` branches._
```

just a suggestion here to be a bit more clear in case someone disputes this

```
▸ src/
  ▸ components/       << The Preact components for the UI of our widgets
  ▸ connectors/       << The source of all the connectors driving the widgets' logic
  ▸ helpers/          << The source of the method helpers
  ▸ lib/              << The core of the library (InstantSearch, routers, etc.)

>   ▸ types/            << The types describing the interfaces

Reviewer: ```suggestion
  ▸ types/            << The TypeScript declarations for global concepts
```

not sure if this is really better

Author: I just went for "The TypeScript declarations" for now.
